sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -111,6 +113,23 @@ object DateTimeUtils {
     computedTimeZones.computeIfAbsent(timeZoneId, computeTimeZone)
   }
 
+  private val threadLocalComputedCalendarsMap =
+    new ThreadLocal[mutable.Map[TimeZone, Calendar]] {
```

Comment:
Do we need to keep Calendar for many timezone? Since `getCalendar` takes a time zone input, we can just keep one Calendar instance, and set it with given timezone in `getCalendar`. WDYT? Regarding performance, is there big difference?

Reply:
Usually, only the default time zone is used. To execute `Cast` regarding date is called with a timezone may use another timezone. For the correctness, I think that it is necessary to support multiple timezones.

To enable caching for default time zone and to create an instance for other time zones would also work correctly.

Reply:
@kiszk I think @viirya meant having just one thread-local calendar instance. That should also work, isn't it?

Reply:
Yes, it should work functionally if we check a given time zone every time. 
Do you know the typical access pattern of time zone? If there is temporal locality regarding time zone, we do not have to use `mutale.Map`.

Reply:
@kiszk @viirya I've tried running benchmarks for with/without ```mutable.Map``` implementation. Looks like setting timezone in a calendar instance is a costly operation and it drags the performance down. As the number of timezones cannot be large, maintaining a map will not be a huge memory overhead. So, I suggest going with the ```mutable.Map``` approach. Comments?

Reply:
How much performance down without a map here?

Reply:
I ran a benchmark and following is the output:
```string to timestamp calendar caching:    Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
with map                                         8 /    9         12.9          77.7       1.0X
without map                                     11 /   12          9.4         106.3       0.7X```

Reply:
+1 for map approach